### PR TITLE
Initial commit for distmv package. Adds Multivariate Gaussian and Mul…

### PR DESCRIPTION
**Pull request by btracey**
…tivariate Uniform distribution

### CODE REVIEW

**Comment by btracey** [comment]
Not read ready for submission, as there are a few TODOs left, but this gives an idea of what I'm thinking.

PTAL @kortschak (and @jonlawlor if you'd like)

File: distmv/general.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package distmv
+
+var (
+	ErrSizeMismatch  = "dist: size mismatch"
```
Comment (kortschak) [inline distmv/general.go]: These aren't errors.

File: distmv/general.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	ErrZeroDimension = "dist: zero dimensional input"
+	ErrReceiver      = "dist: input slice is not nil or the correct length"
+)
+
+const logTwoPi = 1.8378770664093454835606594728112352797227949472755668
+
+// useAs gets a slice of size n. If len(x) == n, x is returned, if x == nil
+// a new slice is allocated, otherwise useAs panics
+func useAs(x []float64, n int) []float64 {
+	if len(x) == n {
```
Comment (kortschak) [inline distmv/general.go]: Do you really want this exact? In mat64 the test is `n < cap(f)` and then it's resliced to n.

File: distmv/general.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+		return x
+	}
+	if x == nil {
```
Comment (kortschak) [inline distmv/general.go]: `len(x) == 0`?

File: distmv/general.go (hunk 1 above)
Comment (btracey) [inline distmv/general.go]: This is how it was done in mat64. What is the appropriate style?

File: distmv/general.go (hunk 2 above)
Comment (kortschak) [inline distmv/general.go]: Ahh, this is equivalent to reuseAs, not use.

File: distmv/general_test.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package distmv
```
Comment (kortschak) [inline distmv/general_test.go]: License.

File: distmv/normal.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/matrix/mat64"
+)
+
+// NormalMV is a multivariate normal distribution (also known as the multivariate
```
Comment (kortschak) [inline distmv/normal.go]: s/NormalMV/Normal/g

File: distmv/normal.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+// Gaussian distribution). It's pdf in k dimensions is given by
```
Comment (kortschak) [inline distmv/normal.go]: s/It's/Its/

File: distmv/uniform.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+// Uniform represents a multivariate uniform distribution
+type Uniform struct {
+	bounds []Bound
+	source *rand.Source
+}
+
+// NewUniform creates a new uniform distribution with the given bounds.
+func NewUniform(bounds []Bound, source *rand.Source) *Uniform {
```
Comment (kortschak) [inline distmv/uniform.go]: s/bounds/b/g
s/source/src/g

File: distmv/general.go (hunk 1 above)
Comment (kortschak) [inline distmv/general.go]: There they are type error, these aren't. If you don't export, then you ca keep it this way.

Comment (btracey) [inline distmv/general.go]: SGTM.

File: distmv/general.go (hunk 3 above)
Comment (btracey) [inline distmv/general.go]: Sure, I guess that's okay.

Comment (kortschak) [inline distmv/general.go]: Doing x = x[:0] is the equivalent to m.Reset().

File: distmv/normal.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	for i := 0; i < dim; i++ {
+		n.logSqrtDet += math.Log(n.chol.At(i, i))
+	}
+	return n, true
+}
+
+// Dim returns the dimension of the distribution.
+func (n *Normal) Dim() int {
+	return len(n.dim)
```
Comment (kortschak) [inline distmv/normal.go]: `return n.dim`

File: distmv/normal_test.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+			t.Errorf("Mean mismatch: Want: %v, got %v", test.mean, estMean)
+		}
+		estCov := stat.CovarianceMatrix(nil, samps, nil)
+		if !estCov.EqualsApprox(cov, 1e-2) {
+			t.Errorf("Cov mismatch: Want: %v, got %v", cov, estCov)
+		}
+	}
+}
+
+func expandMat(data [][]float64) []float64 {
```
Comment (kortschak) [inline distmv/normal_test.go]: Can't we just use line breaks to do this?

File: distmv/uniform.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	for _, b := range bounds {
+		u.bounds[i].Min = b.Min
+		u.bounds[i].Max = b.Max
+	}
+}
+
+// Dim returns the dimension of the distribution.
+func (u *Uniform) Dim() int {
+	return len(u.min)
```
Comment (kortschak) [inline distmv/uniform.go]: u.min does not exist.

File: distmv/uniform.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+// LogProb computes the log of the pdf of the point x.
+func (u *Uniform) LogProb(x []float64) float64 {
+	dim := len(u.min)
```
Comment (kortschak) [inline distmv/uniform.go]: u.min does not exist.

File: distmv/uniform.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+// Prob computes the value of the probability density function at x.
+func (u *Uniform) Prob(x []float64) float64 {
+	return math.Exp(n.LogProb(x))
+}
+
+// Rand generates a random number according to the distributon.
+// If the input slice is nil, new memory is allocated, otherwise the result is stored
+// in place.
+func (u *Uniform) Rand(x []float64) []float64 {
+	x = useAs(x, n.dim)
```
Comment (kortschak) [inline distmv/uniform.go]: n should be u and u.dim does not exist.

File: distmv/uniform.go [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+		logprob -= math.Log(b.Max - b.Min)
+	}
+	return logprob
+}
+
+// Mean returns the mean of the probability distribution at x. If the
+// input argument is nil, a new slice will be allocated, otherwise the result
+// will be put in-place into the receiver.
+func (u *Uniform) Mean(x []float64) []float64 {
+	x = useAs(x, n.dim)
```
Comment (kortschak) [inline distmv/uniform.go]: n should be u and u.dim does not exist.

File: distmv/uniform.go [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	u := &Uniform{
+		bounds: make([]Bound, len(bounds)),
+		source: source,
+	}
+	for _, b := range bounds {
+		u.bounds[i].Min = b.Min
+		u.bounds[i].Max = b.Max
+	}
+}
```
Comment (kortschak) [inline distmv/uniform.go]: `return u`.

File: distmv/uniform.go [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+	return logprob
+}
+
+// Mean returns the mean of the probability distribution at x. If the
+// input argument is nil, a new slice will be allocated, otherwise the result
+// will be put in-place into the receiver.
+func (u *Uniform) Mean(x []float64) []float64 {
+	x = useAs(x, n.dim)
+	for _, b := range u.bounds {
+		x[i] = (b.Max + b.Min) / 2
```
Comment (kortschak) [inline distmv/uniform.go]: s/_/i/

File: distmv/normal.go [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+// Entropy returns the differential entropy of the distribution.
+func (n *Normal) Entropy() float64 {
+	return float64(dim)/2*(1+logTwoPi) + n.logSqrtDet
```
Comment (kortschak) [inline distmv/normal.go]: s/dim/n.dim/

File: distmv/normal.go (hunk 8 above)
Comment (btracey) [inline distmv/normal.go]: Ugh. Sorry.

File: distmv/general.go (hunk 3 above)
Comment (btracey) [inline distmv/general.go]: Done.

File: distmv/general_test.go (hunk 4 above)
Comment (btracey) [inline distmv/general_test.go]: Done.

File: distmv/normal.go (hunk 5 above)
Comment (btracey) [inline distmv/normal.go]: Done.

File: distmv/normal.go (hunk 6 above)
Comment (btracey) [inline distmv/normal.go]: Done.

File: distmv/uniform.go (hunk 7 above)
Comment (btracey) [inline distmv/uniform.go]: Done.

File: distmv/normal_test.go (hunk 9 above)
Comment (btracey) [inline distmv/normal_test.go]: Done.

**Comment by btracey** [comment]
PTAL

File: distmv/general_test.go [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+	dist    prober
+	loc     []float64
+	logProb float64
+}
+
+func testProbability(t *testing.T, cases []probCase) {
+	for _, test := range cases {
+		logProb := test.dist.LogProb(test.loc)
+		if math.Abs(logProb-test.logProb) > 1e-14 {
+			t.Errorf("LogProb mismatch: Want: %v, got: %v", test.logProb, logProb)
```
Comment (kortschak) [inline distmv/general_test.go]: lc/Want/

and below.

**Comment by kortschak** [comment]
The code LGTM, but it would be good to have someone with stronger stats also say OK. Failing that... SHIP IT!

**Comment by btracey** [comment]
Are the concerns about correctness or API?

**Comment by kortschak** [comment]
Correctness. I'm sure it's fine, but MVA is not a strong point of mine.

**Comment by kortschak** [comment]
@Armadilloa16 would you have a look at this?

File: distmv/general_test.go (hunk 17 above)
Comment (btracey) [inline distmv/general_test.go]: What is "Ic"?

Comment (btracey) [inline distmv/general_test.go]: oh. lowercase.

**Comment by Armadilloa16** [comment]
Seems to make sense, although I would need to be more familiar with the go language and spend more time checking it thoroughly in order to be confident it is correct. From just looking it over quickly the concept seems sound, particularly as a starting point that can be expanded on.

**Comment by btracey** [comment]
Thanks.

**Comment by kortschak** [comment]
@btracey (Brendan), meet @Armadilloa16 (Lyron).

**Comment by btracey** [comment]
Nice to meet you. Hope we'll be seeing you around here more!

**Comment by Armadilloa16** [comment]
Hey nice to meet you too @btracey! I hope to be around more! I'm new to the whole software collaboration thing, but keen to learn and contribute. Was thinking of doing some stats as thats my background, maybe some classification stuff. Just a question of how much free time I can scrounge up haha. :)

**Comment by jonlawlor** [comment]
That's usually the issue!

Also, I was looking at `math\rand\rand.go` and I noticed that the default source is safe for concurrent use, but `Rand.New()` is not.  It might be a good idea to document that the distribution objects safety is dependent on the Rand.Source object  - I was not expecting it.

**Comment by btracey** [comment]
If you can find an appropriate place to comment, then sure, but it's not really our problem. None of our API call random numbers concurrently for this reason.

**Comment by btracey** [comment]
Oh, I see. Your point is if they set the Source and then call the distribution concurrently.

**Comment by kortschak** [comment]
In general types are not safe for concurrent use unless specified as such. I don't think we need to say anything more than "X is safe for concurrent use if src is safe for concurrent use." at NewX.

**Comment by btracey** [comment]
Probably better as a package comment since it's a general property of dist and distmv.

**Comment by kortschak** [comment]
Sure.

**Comment by jonlawlor** [comment]
The tricky part is that it is safe for concurrent use, unless you specify your own source.  I can easily imagine a scenario where someone is testing out a MCMC simulation with the default Source (a problem that is embarrassingly parallelizable), then wants to "do things right" and uses a defined random seed, and ends up with something that is no longer safe.  I agree, a package comment is appropriate.

**Comment by btracey** [comment]
You can set a random seed and still be safe for concurrent use (by setting rand.Seed(myint)). 

The issue with rand.Float64() is that it does not scale in parallel if you need a lot of random numbers. The way to fix that is to generate several *rand.Rand, and have a separate one in each concurrent thread. A well-written program will also be safe for concurrent use, since the point was to eliminate contention. Alternately, rand.Source is an interface, so the Int63() method could be designed for safe concurrent use.

**Pull request merged**